src/pubsub/node_flat.erl

```diff
@@ -525,52 +532,36 @@ set_subscriptions(Nidx, Owner, Subscription, SubId) ->
                         ?ERR_EXTENDED((mongoose_xmpp_errors:bad_request()), <<"not-subscribed">>)};
                 _ ->
                     NewSubId = pubsub_subscription:make_subid(),
-                    mod_pubsub_db_backend:add_subscription(Nidx, Owner, Subscription, NewSubId)
+                    mod_pubsub_db_backend:add_subscription(Nidx, LOwner, Subscription, NewSubId)
             end;
         {<<>>, [{_, SID}]} ->
             case Subscription of
-                none -> mod_pubsub_db_backend:delete_subscription(Nidx, Owner, SID);
-                _ -> mod_pubsub_db_backend:update_subscription(Nidx, Owner, Subscription, SID)
+                none -> mod_pubsub_db_backend:delete_subscription(Nidx, LOwner, SID);
+                _ -> mod_pubsub_db_backend:update_subscription(Nidx, LOwner, Subscription, SID)
             end;
         {<<>>, [_ | _]} ->
             {error,
                 ?ERR_EXTENDED((mongoose_xmpp_errors:bad_request()), <<"subid-required">>)};
         _ ->
             case Subscription of
-                none -> mod_pubsub_db_backend:delete_subscription(Nidx, Owner, SubId);
-                _ -> mod_pubsub_db_backend:update_subscription(Nidx, Owner, Subscription, SubId)
+                none -> mod_pubsub_db_backend:delete_subscription(Nidx, LOwner, SubId);
+                _ -> mod_pubsub_db_backend:update_subscription(Nidx, LOwner, Subscription, SubId)
             end
     end.
 
 %% @doc <p>Returns a list of Owner's nodes on Host with pending
 %% subscriptions.</p>
 get_pending_nodes(Host, Owner) ->
-    {ok, States} = mod_pubsub_db_backend:get_own_nodes_states(Owner),
+    LOwner = jid:to_lower(Owner),
+    {ok, Nidxs} = mod_pubsub_db_backend:get_idxs_of_own_nodes_with_pending_subs(LOwner),
     NodeTree = mod_pubsub:tree(Host),
-    Reply = lists:foldl(fun (PubSubState, Acc) ->
-                                case get_node_if_has_pending_subs(NodeTree, PubSubState) of
-                                    {value, Node} -> [Node | Acc];
-                                    false -> Acc
-                                end
-                         end,
-                         [], States),
-    {result, Reply}.
-
-get_node_if_has_pending_subs(NodeTree, #pubsub_state{stateid = {_, N}, subscriptions = Subs}) ->
```

Comment:
Where did the function go?

Reply:
The filtering for pending state has been moved to backend (so it may be much more efficient for RDBMS).